File: eppo/src/main/java/cloud/eppo/android/dto/PrecomputedBandit.java
package cloud.eppo.android.dto;

import androidx.annotation.Nullable;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Map;

/**
 * Represents a precomputed bandit assignment from the edge endpoint. String fields are Base64
 * encoded.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class PrecomputedBandit {

  private final String banditKey;
  private final String action;
  private final String modelVersion;
  @Nullable private final Map<String, String> actionNumericAttributes;
  @Nullable private final Map<String, String> actionCategoricalAttributes;
  private final double actionProbability;
  private final double optimalityGap;

  @JsonCreator
  public PrecomputedBandit(
      @JsonProperty("banditKey") String banditKey,
      @JsonProperty("action") String action,
      @JsonProperty("modelVersion") String modelVersion,
      @JsonProperty("actionNumericAttributes") @Nullable Map<String, String> actionNumericAttributes,
      @JsonProperty("actionCategoricalAttributes") @Nullable Map<String, String> actionCategoricalAttributes,
      @JsonProperty("actionProbability") double actionProbability,
      @JsonProperty("optimalityGap") double optimalityGap) {
    this.banditKey = banditKey;
    this.action = action;
    this.modelVersion = modelVersion;
    this.actionNumericAttributes = actionNumericAttributes;
    this.actionCategoricalAttributes = actionCategoricalAttributes;
    this.actionProbability = actionProbability;
    this.optimalityGap = optimalityGap;
  }

  /** Returns the Base64-encoded bandit key. */
  public String getBanditKey() {
    return banditKey;
  }

  /** Returns the Base64-encoded action. */
  public String getAction() {
    return action;
  }

  /** Returns the Base64-encoded model version. */
  public String getModelVersion() {
    return modelVersion;
  }

  /** Returns the Base64-encoded numeric attributes for the action. */
  @Nullable public Map<String, String> getActionNumericAttributes() {
    return actionNumericAttributes;
  }

  /** Returns the Base64-encoded categorical attributes for the action. */
  @Nullable public Map<String, String> getActionCategoricalAttributes() {
    return actionCategoricalAttributes;
  }

  /** Returns the probability of taking this action. */
  public double getActionProbability() {
    return actionProbability;
  }

  /** Returns the gap to the optimal action. */
  public double getOptimalityGap() {
    return optimalityGap;
  }
}

File: eppo-kotlin/src/main/kotlin/cloud/eppo/kotlin/internal/network/dto/PrecomputedBandit.kt
package cloud.eppo.kotlin.internal.network.dto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Precomputed bandit data from the server.
 *
 * Key string fields (banditKey, action, modelVersion) are Base64 encoded.
 * Attribute maps have Base64 encoded keys and values.
 *
 * @property banditKey Base64 encoded bandit identifier
 * @property action Base64 encoded recommended action
 * @property modelVersion Base64 encoded model version
 * @property actionProbability Probability of this action being selected
 * @property optimalityGap Optimality gap metric
 * @property actionNumericAttributes Numeric attributes (Base64 encoded keys/values)
 * @property actionCategoricalAttributes Categorical attributes (Base64 encoded keys/values)
 */
@Serializable
internal data class PrecomputedBandit(
    @SerialName("banditKey") val banditKey: String,
    @SerialName("action") val action: String,
    @SerialName("modelVersion") val modelVersion: String,
    @SerialName("actionProbability") val actionProbability: Double,
    @SerialName("optimalityGap") val optimalityGap: Double,
    @SerialName("actionNumericAttributes") val actionNumericAttributes: Map<String, String> = emptyMap(),
    @SerialName("actionCategoricalAttributes") val actionCategoricalAttributes: Map<String, String> = emptyMap()
)
